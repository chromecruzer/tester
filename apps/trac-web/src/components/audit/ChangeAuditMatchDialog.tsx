import {DialogModal} from "../tables/DialogModal";
import {
  AuditMatch,
  auditMatch, dataTableNames,
  NullableString,
} from "@trac/datatypes";
import React, {useEffect, useState} from "react";
import {MdOutlineClear} from "react-icons/md";
import {SearchWithAutoComplete} from "../search/SearchWithAutoComplete";
import {useMutation, useQueryClient} from "@tanstack/react-query";
import {getAudit, postAuditNotes, postBatchPriceUpdate} from "../../api/auditApi";
import {useTracContext} from "../../TracContext";

export const initialLocationSignature = {
  uuid: '',
  customer_code: '',
  warehouse: '',
}

export const ChangeAuditMatchDialog = ({batchEdit, open, requestClose, auditUuid}) => {
  const [note, setNote] = useState('');
  const [match, setMatch] = useState('Other' as AuditMatch);
  const [location, setLocation] = useState(initialLocationSignature);
  const [locationChosen, setLocationChosen] = useState(false);
  const [error, setError] = useState(null as NullableString);
  const [isLoading, setIsLoading] = useState(false);
  const queryClient = useQueryClient();
  const {setTracContext} = useTracContext();
  useEffect(() => {
    const filters = {};
    filters[dataTableNames.customers] = true;
    setTracContext('search', {filters})
  }, [])
  const postAuditNotesMutation = useMutation({
    mutationFn: async () => {
      setIsLoading(true);
      setError(null);
      handleClose();
      return postAuditNotes(batchEdit.createItemNote(`${note}\nMatch changed to ${match}`,
        'Match'));
    },
    onSuccess: (data, variables, context) => {
      setError(null);
      queryClient.prefetchQuery(['audit'], () => getAudit(auditUuid));
    },
    onError: (error: NullableString, variables, context) => {
      if (error) {
        setError(error);
        console.error('post audit notes', error);
      }
    },
    onSettled: (data, error, variables, context) => {
      setIsLoading(false);
    }
  });

  const postBatchChangeMutation = useMutation({
    mutationFn: async () => {
      setIsLoading(true);
      setError(null);
      const batch = match === 'Found In Other Location' ?
        batchEdit.moved(location.uuid, location.customer_code, location.warehouse) :
        batchEdit.matchChange(match)
      handleClose();
      return postBatchPriceUpdate(batch);
    },
    onSuccess: (data, variables, context) => {
      setError(null);
      queryClient.invalidateQueries({queryKey: ['audit', auditUuid]})
    },
    onError: (error: NullableString, variables, context) => {
      if (error) {
        setError(error);
        console.error('post audit notes', error);
      }
    },
    onSettled: (data, error, variables, context) => {
      setIsLoading(false);
      handleClose();
    }
  })
  const handleClose = () => {
    setNote('');
    requestClose();
  }
  const choose = (choice) => {
    setLocationChosen(true);
    setLocation({
      ...location,
      customer_code: choice.suggestion.signature.customer_code,
      uuid: choice.suggestion.uuid
    })
  }

  const handleWarehouse = (warehouse: string) => {
    setLocation({...location, warehouse});
  }

  const locationField = locationChosen ?
    <div className='flex justify-start items-start border-2 border-bl-consumer-off-white'>
      <h4 className='py-2 px-3'>Account:</h4>
      <div className='flex border-2 hover: border-bl-consumer-dark'>
        <span className='py-2 px-3 text-bl-text-grey'>{location.customer_code}</span>
        <button onClick={() => setLocationChosen(false)}><MdOutlineClear/></button>
      </div>
    </div> :
    <SearchWithAutoComplete
      setChosen={choose}
    >
    </SearchWithAutoComplete>
  const movedFields = match === 'Found In Other Location' ?
    <div className='flex flex-row'>
      {locationField}
      <div className='flex'>
        <h4 className='py-2 px-3'>Warehouse:</h4>
        <input
          type='text'
          value={location.warehouse || ''}
          onChange={e => handleWarehouse(e.target.value)}
          className='py-2 px-3 flex-grow cursor-pointer bg-bl-text-light text-bl-text-grey border-2 border-bl-consumer-off-white'
        />
      </div>
    </div>
    : null;

  const handleChange = e => {
    setNote(e.target.value);
  }

  const handleNewMatch = e => {
    setMatch(e.target.value);
  }

  const handleSubmit = () => {
    console.log(`change match ${match} with note ${note}`)
    postBatchChangeMutation.mutate();
    postAuditNotesMutation.mutate();
  }

  return <DialogModal open={open} requestClose={handleClose}>
    <div className='flex-col min-h-full min-w-full justify-items-stretch'>
      <div className='flex flex-col justify-start items-center'>
        <div className='flex'>
          <h4 className='py-2 px-3 min-w-0'>Match Type</h4>
          <select
            value={match}
            onChange={handleNewMatch}
            className='py-2 px-3 uppercase cursor-pointer bg-bl-text-light text-bl-text-grey'
          >
            {auditMatch.map(k => (<option key={k}>{k}</option>))}
          </select>
        </div>
        {movedFields}
      </div>
      <h2 className='px-2 py-2 text-center font-bold text-bl-text-dark'>Note</h2>
      <textarea
        value={note}
        rows={8}
        placeholder='Type your note here'
        onChange={handleChange}
        className='border-2 px-2 py-2 grow justify-self-stretch self-stretch min-h-0 min-w-0 w-full h-full'
      />
      <div className='flex-row content-between'>
        <button className='border-2 px-2 py-2 filter-button-active' onClick={handleClose}>Cancel</button>
        <button className='border-2 px-2 py-2 filter-button-active' onClick={handleSubmit}>Ok</button>
      </div>
    </div>
  </DialogModal>
}
