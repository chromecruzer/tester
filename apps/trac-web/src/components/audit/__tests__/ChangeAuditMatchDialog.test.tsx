import {render, screen} from "@testing-library/react";
import {ChangeAuditMatchDialog, initialLocationSignature, LocationSignature} from "../ChangeAuditMatchDialog";
import {useState as MockUseState} from "react";
import * as MockHooks from "../../../redux/hooks";
import * as MockAuditActions from "../../../redux/auditActions";
import userEvent from "@testing-library/user-event";

jest.mock('react', () => ({
  ... jest.requireActual('react'),
  useState: jest.fn()
}))
jest.mock('../../tables/DialogModal', () => ({
  DialogModal: ({open, requestClose, children}) => (
    <div>
      <h4>open is set to {open}</h4>
      <h4>requestClose is {requestClose ? 'present' : 'absent'}</h4>
      {children}
    </div>)
}))
jest.mock('../../location/LocationAutoComplete', () => ({
  LocationAutoComplete: ({page, locationUuid, returnChosen}) => (
    <div>
      <h4>page is set to {page}</h4>
      <h4>locationUuid is set to {locationUuid}</h4>
      <h4>returnChosen is {returnChosen ? 'present' : 'absent'}</h4>
      <button onClick={() => returnChosen({
        field: 'field 1',
        uuid: 'customer uuid',
        type: 'customer',
        signature: {
          uuid: 'location uuid',
          customer_code: 'location code',
          name: 'West Mall',
          address: '123 Mickey Mouse Way',
          city:'Orlando',
          state: 'Florida',
          postal: '123456'
        }
      })}>Choose Location</button>
    </div>)
}))
describe('ChangeAuditMatchDialog', () => {
  let mockedBatch, closeSpy, dispatchSpy, postAuditNotesSpy, postBatchItemAuditUpdateSpy, getAuditItemsSpy,
    setNoteSpy, setMatchSpy, setLocationSpy, setLocationChoseSpy;
  const renderChangeMatchDialog = () => {
    render(<ChangeAuditMatchDialog batchEdit={mockedBatch} open={true} requestClose={closeSpy} auditUuid={'audit uuid'}/>)
  }
  const useStateReturns = (content='', match, location = null as (null | LocationSignature)) => {
    (MockUseState as jest.Mock)
      .mockImplementationOnce(() => [content, setNoteSpy])
      .mockImplementationOnce(() => [match, setMatchSpy])
      .mockImplementationOnce(() => [location || initialLocationSignature, setLocationSpy])
      .mockImplementationOnce(() => [location !== null, setLocationChoseSpy]);
  }
  beforeEach(() => {
    mockedBatch = {
      createItemNote: jest.fn(),
      matchChange: jest.fn(),
      moved: jest.fn()
    };
    closeSpy = jest.fn();
    dispatchSpy = jest.fn();
    setNoteSpy = jest.fn();
    setMatchSpy = jest.fn();
    setLocationChoseSpy = jest.fn();
    setLocationSpy = jest.fn();
    jest.spyOn(MockHooks, 'useAppDispatch').mockImplementation(() => dispatchSpy);
    postAuditNotesSpy = jest.spyOn(MockAuditActions, 'postAuditNotes').mockReturnValue(() => Promise.resolve());
    postBatchItemAuditUpdateSpy = jest.spyOn(MockAuditActions, 'postBatchItemAuditUpdate').mockReturnValue(() => Promise.resolve());
    getAuditItemsSpy = jest.spyOn(MockAuditActions, 'getAudit').mockReturnValue(() => Promise.resolve());
  });
  it('should save a typed note', () => {
    const expected = 'Adding a note\nAdios!';
    userEvent.setup();
    useStateReturns('Adding a note\nAdios', 'Other');
    renderChangeMatchDialog();
    const textArea = screen.getByRole('textbox');
    return userEvent.type(textArea, '!')
      .then(() => {
        // console.log('typing was done');
        expect(setNoteSpy).toBeCalledWith(expected);
      });
  });
  it('should save a match change', () => {
    userEvent.setup();
    useStateReturns('Adding a note\nAdios', 'Other');
    renderChangeMatchDialog();
    const select = screen.getByRole('combobox');
    return userEvent.selectOptions(select, 'Billed').then(() => {
      expect(setMatchSpy).toBeCalledWith('Billed');
    });
  });
  it('should should submit a match change', () => {
    const expected = 'Recently typed note';
    userEvent.setup();
    useStateReturns(expected, 'Billed');
    renderChangeMatchDialog();
    const button = screen.getByRole('button', {name: 'Ok'});
    return userEvent.click(button)
      .then(() => {
        // console.log('OK was clicked');
        expect(postAuditNotesSpy).toBeCalled();
        expect(postBatchItemAuditUpdateSpy).toBeCalled();
        expect(dispatchSpy).toBeCalled();
        expect(mockedBatch.createItemNote.mock.calls).toMatchSnapshot();
        expect(mockedBatch.matchChange).toBeCalledWith('Billed');
        expect(getAuditItemsSpy).toBeCalled();
        expect(closeSpy).toBeCalled();
      })
  });
  it('should display location field when match is "moved"', () => {
    const expected = 'Recently typed note';
    userEvent.setup();
    useStateReturns(expected, 'Found In Other Location');
    renderChangeMatchDialog();
    screen.getByText('page is set to Audit');
    screen.getByText('returnChosen is present');
  });
  it('should save a location', () => {
    const expected = 'Recently typed note';
    userEvent.setup();
    useStateReturns(expected, 'Found In Other Location');
    renderChangeMatchDialog();
    const button = screen.getByRole('button', {name: 'Choose Location'});
    // screen.debug(button)
    return userEvent.click(button)
      .then(() => {
        // console.log('chosen button clicked');
        expect(setLocationChoseSpy).toBeCalledWith(true);
        expect(setLocationSpy.mock.calls).toMatchSnapshot();
      })
  });
  it('should should submit a location with a match changed to "moved"', () => {
    const expected = 'Recently typed note';
    userEvent.setup();
    useStateReturns(expected, 'Found In Other Location', {
      uuid: 'location uuid',
      customer_code: 'location code',
      name: 'West Mall',
      address: '123 Mickey Mouse Way',
      city:'Orlando',
      state: 'Florida',
      postal: '123456'
    });
    renderChangeMatchDialog();
    const button = screen.getByRole('button', {name: 'Ok'});
    return userEvent.click(button)
      .then(() => {
        // console.log('OK was clicked');
        expect(postAuditNotesSpy).toBeCalled();
        expect(postBatchItemAuditUpdateSpy).toBeCalled();
        expect(dispatchSpy).toBeCalled();
        expect(mockedBatch.createItemNote.mock.calls).toMatchSnapshot();
        expect(mockedBatch.moved.mock.calls).toMatchSnapshot();
        expect(getAuditItemsSpy).toBeCalled();
        expect(closeSpy).toBeCalled();
      })
  });
});
