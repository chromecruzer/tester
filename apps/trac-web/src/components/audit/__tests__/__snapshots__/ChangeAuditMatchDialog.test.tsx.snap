// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeAuditMatchDialog should save a location 1`] = `
Array [
  Array [
    Object {
      "address": "123 Mickey Mouse Way",
      "city": "Orlando",
      "customer_code": "location code",
      "name": "West Mall",
      "postal": "123456",
      "state": "Florida",
      "uuid": "customer uuid",
    },
  ],
]
`;

exports[`ChangeAuditMatchDialog should should submit a location with a match changed to "moved" 1`] = `
Array [
  Array [
    "Recently typed note
Match changed to Found In Other Location",
    "Match",
  ],
]
`;

exports[`ChangeAuditMatchDialog should should submit a location with a match changed to "moved" 2`] = `
Array [
  Array [
    "location uuid",
    "location code",
  ],
]
`;

exports[`ChangeAuditMatchDialog should should submit a match change 1`] = `
Array [
  Array [
    "Recently typed note
Match changed to Billed",
    "Match",
  ],
]
`;
